src/virt.rs

```diff
@@ -1001,3 +1001,67 @@ where
         self.set_csr(*register, value, hw)
     }
 }
+
+// ————————————————————————————————— Tests —————————————————————————————————— //
+
+#[cfg(test)]
+mod tests {
+    use crate::arch::{mstatus, Arch, Architecture, Csr, Mode};
+    use crate::host::MiralisContext;
+    use crate::virt::VirtContext;
+
+    #[test]
+    fn switch_context_mpp() {
+        let hw = unsafe { Arch::detect_hardware() };
+        let mut mctx = MiralisContext::new(16, hw);
+        let mut ctx = VirtContext::new(0, 16);
+
+        ctx.csr.mstatus |= Mode::S.to_bits() << mstatus::MPP_OFFSET;
+
+        unsafe { ctx.switch_from_firmware_to_payload(&mut mctx) }
+
+        assert_eq!(
+            ctx.csr.mstatus & mstatus::MPP_FILTER,
+            Mode::S.to_bits() << mstatus::MPP_OFFSET,
+            "VirtContext Mstatus.MPP need to be set to S mode (going to payload)"
+        );
+
+        assert_eq!(
+            Arch::read_csr(Csr::Mstatus) & mstatus::MPP_FILTER,
+            Mode::M.to_bits() << mstatus::MPP_OFFSET,
+            "Mstatus.MPP need to be set to M mode (coming from Miralis)"
+        );
+
+        // Simulate a trap
+        unsafe { Arch::write_csr(Csr::Mstatus, Mode::S.to_bits() << mstatus::MPP_OFFSET) };
+
+        unsafe { ctx.switch_from_payload_to_firmware(&mut mctx) }
+
+        // VirtContext Mstatus.MPP has been set to M mode
+        assert_eq!(
+            ctx.csr.mstatus & mstatus::MPP_FILTER,
+            Mode::S.to_bits() << mstatus::MPP_OFFSET,
+            "VirtContext Mstatus.MPP has been set to S mode (coming from payload)"
+        );
+
+        // Mstatus.MPP has been set to U mode
+        assert_eq!(
+            Arch::read_csr(Csr::Mstatus) & mstatus::MPP_FILTER,
+            Mode::U.to_bits() << mstatus::MPP_OFFSET,
+            "Mstatus.MPP has been set to U mode (going to firmware)"
+        );
+    }
+
+    #[test]
+    fn switch_to_firmware_mideleg() {
+        let hw = unsafe { Arch::detect_hardware() };
+        let mut mctx = MiralisContext::new(16, hw);
+        let mut ctx = VirtContext::new(0, 16);
+
+        unsafe { Arch::write_csr(Csr::Mideleg, 0x42) };
+
+        unsafe { ctx.switch_from_payload_to_firmware(&mut mctx) }
+
+        assert_eq!(Arch::read_csr(Csr::Mideleg), 0, "Mideleg need to be 0");
+    }
+}
```

Comment:
It's great to have tests, but we should also think about when we will have to figure out why they fail ^^

As it is not immediately obvious what the tests are doing, could you add a doc comment for each of those functions to express what is the intent of the test? What part are you trying to test, how is the system expected to behave, and what are the expected invariants/properties?

Reply:
Of course, I'll add this !